### PR TITLE
gh-90953: Emit deprecation warnings for deprecated ast features

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
I am not sure that it is worth to include it in 3.11. It may be better to wait for 3.12 or 3.13.

https://bugs.python.org/issue46797

* Issue: gh-90953

### CODE REVIEW

**Comment by AlexWaygood** [comment]
Could we hold off doing this until 3.7 is EOL? It is quite hard to write version-compatible code otherwise. (I'm speaking as a maintainer of https://github.com/PyCQA/flake8-pyi.)

**Comment by serhiy-storchaka** [comment]
Yes, this is my intention.

**Comment by serhiy-storchaka** [comment]
Well, I am going to merge it.

Note, that it does not remove previously deprecated feature. It only adds runtime warnings for already long deprecated (3 or more releases ago) features. It is much much more smoother than standard practice of adding runtime warnings at the moment of deprecation.

**Comment by AlexWaygood** [comment]
I support merging this, but I would still leave the deprecation warning for `ExtSlice` to a future PR, personally. Unlike the other deprecated features, it was deprecated in 3.9, not 3.7.

**Comment by hugovk** [comment]
(Merge conflict resolved)

There's about a week until the 3.12 cutoff. I think it's okay to include `ExtSlice` here, it's replacement `Tuple` has been around for a long time, and `ast.*ExtSlice` shows up in 22 projects in the PyPI top 5k:

```console
❯ python3 ~/github/misc/cpython/search_pypi_top.py -q . "ast.*ExtSlice"
./pyre-check-nightly-0.0.101680433836.tar.gz: pyre-check-nightly-0.0.101680433836/typeshed/stubs/pyflakes/pyflakes/checker.pyi: def EXTSLICE(self, tree: ast.ExtSlice, omit: _OmitType = ...) -> None: ...
./reportlab-3.6.12.tar.gz: reportlab-3.6.12/src/reportlab/lib/rl_safe_eval.py: elif isinstance(slice_, ast.ExtSlice):
./typed_ast-1.5.4.tar.gz: typed_ast-1.5.4/ast27/Python/Python-ast.c: value = ast2obj_list(o->v.ExtSlice.dims, ast2obj_slice);
./typed_ast-1.5.4.tar.gz: typed_ast-1.5.4/ast3/Python/Python-ast.c: value = ast2obj_list(o->v.ExtSlice.dims, ast2obj_slice);
./mypy-1.1.1.tar.gz: mypy-1.1.1/mypy/fastparse.py: is_py38_or_earlier and isinstance(n.slice, ast3.ExtSlice)
./mypy-1.1.1.tar.gz: mypy-1.1.1/mypy/fastparse.py: def visit_ExtSlice(self, n: ast3.ExtSlice) -> TupleExpr:
./mypy-1.1.1.tar.gz: mypy-1.1.1/mypy/fastparse.py: assert isinstance(n.slice, ast3.ExtSlice)
./pure_eval-0.2.2.tar.gz: pure_eval-0.2.2/pure_eval/core.py: elif isinstance(index, ast.ExtSlice):
./gast-0.5.3.tar.gz: gast-0.5.3/gast/ast3.py: new_slice = ast.ExtSlice(
./gast-0.5.3.tar.gz: gast-0.5.3/gast/ast2.py: new_slice = ast.ExtSlice([adjust_slice(self._visit(elt))
./ipython-8.12.0.tar.gz: ipython-8.12.0/IPython/core/guarded_eval.py: if isinstance(node, ast.ExtSlice):
./flake8_expression_complexity-0.0.11.tar.gz: flake8_expression_complexity-0.0.11/flake8_expression_complexity/utils/complexity.py: (ast.ExtSlice, 'ext_slice'),
./astroid-2.15.1.tar.gz: astroid-2.15.1/astroid/nodes/node_classes.py: """Class representing an :class:`ast.ExtSlice` node.
./astroid-2.15.1.tar.gz: astroid-2.15.1/astroid/rebuilder.py: def visit(self, node: ast.ExtSlice, parent: nodes.Subscript) -> nodes.Tuple:
./astroid-2.15.1.tar.gz: astroid-2.15.1/astroid/rebuilder.py: self, node: ast.ExtSlice, parent: nodes.Subscript
./asttokens-2.2.1.tar.gz: asttokens-2.2.1/asttokens/asttokens.py: ast.Slice, ast.ExtSlice, ast.Index, ast.keyword,
./asttokens-2.2.1.tar.gz: asttokens-2.2.1/asttokens/asttokens.py: - ``ast.ExtSlice``
./griffe-0.25.5.tar.gz: griffe-0.25.5/src/griffe/agents/nodes.py: from ast import ExtSlice as NodeExtSlice
./numba-0.56.4.tar.gz: numba-0.56.4/numba/tests/test_stencils.py: slice=ast.ExtSlice(
./numba-0.56.4.tar.gz: numba-0.56.4/numba/tests/test_stencils.py: if isinstance(node.slice, ast.ExtSlice):
./pytype-2023.3.31.tar.gz: pytype-2023.3.31/pytype/typeshed/stubs/pyflakes/pyflakes/checker.pyi: def EXTSLICE(self, tree: ast.ExtSlice, omit: _OmitType = ...) -> None: ...
./RestrictedPython-6.0.tar.gz: RestrictedPython-6.0/src/RestrictedPython/transformer.py: elif isinstance(slice_, ast.ExtSlice):

Time: 0:00:17.604108
Found 22 matching lines in 14 projects
```

But I'm also to okay to splitting it into another PR and merging it after the 3.13 branch in about a week and a few days.

In any case, let's add this to What's New.

File: Lib/test/test_ast.py [hunk 1]
```diff
@@ -435,13 +437,48 @@ def test_base_classes(self):
...
+                     'slice', 'Index', 'ExtSlice'):
+            item = getattr(ast, name)
             if self._is_ast_node(name, item):
                 if name == 'Index':
                     # Index(value) just returns value now.
                     # The argument is required.
                     continue
+                with self.subTest(item):
+                 with self.assertWarns(DeprecationWarning):
+                    x = item()
```
Comment (hugovk) [inline Lib/test/test_ast.py]: Fix `make patchcheck`:
```suggestion
                with self.subTest(item):
                    with self.assertWarns(DeprecationWarning):
                        x = item()
```

**Comment by hugovk** [comment]
A suggested What's New entry:

```rst
* The following :mod:`ast` features have been deprecated in documentation since
  Python 3.8 and now raise a proper :exc:`DeprecationWarning`:
  :class:`!ast.Num`, :class:`!ast.Str`, :class:`!ast.Bytes`,
  :class:`!ast.NameConstant` and :class:`!ast.Ellipsis`; use :class:`ast.Constant`
  instead.
  Similarly since 3.9, instead of :class:`!ast.Index` use the index value
  directly; and instead of :class:`!ast.ExtSlice` use :class:`ast.Tuple`.
```

**Comment by AlexWaygood** [comment]
Well, my position remains the same: I'd love to see this get into 3.12, but I'd really rather we excluded `ExtSlice` and `Index` from this PR for now. It's not so much a matter of finding a _replacement_ for the class -- the issue is that it will complicate life for people writing linters using `ast.NodeVisitor` subclasses to inspect Python source code, if you want to support Python 3.8-3.12, and you want to avoid `DeprecationWarning`s.

**Comment by AlexWaygood** [comment]
> Okay, would you like to create a new PR with just the 3.8 ones and we can merge that into 3.12 before the cutoff?

Sounds good. I'll list @serhiy-storchaka as a co-author.

**Comment by AlexWaygood** [comment]
(Accidentally pushed to this PR instead of #104199. Force-pushed to undo that.)

**Comment by furkanonder** [comment]
cc: @isidentical

**Comment by serhiy-storchaka** [comment]
Closed in favor of #104199.

@AlexWaygood, do you mind to create a new PR for ExtSlice and Index?

**Comment by AlexWaygood** [comment]
> Closed in favor of #104199.
> 
> @AlexWaygood, do you mind to create a new PR for ExtSlice and Index?

Sure, I'm working on it now

**Pull request closed without merging**